src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+                            break :result {};
+                        }
+                    };
+                    completion.callback(completion.context, completion, &result);
+                },
+                .openat => |op| {
+                    const result = result: {
+                        if (completion.result < 0) {
+                            break :result switch (-completion.result) {
+                                os.EINTR => {
+                                    completion.prep();
+                                    return;
+                                },
+                                os.EACCES => error.AccessDenied,
+                                os.EBADF => error.FileDescriptorInvalid,
+                                os.EBUSY => error.DeviceBusy,
+                                os.EEXIST => error.PathAlreadyExists,
+                                os.EFAULT => unreachable,
+                                os.EFBIG => error.FileTooBig,
+                                os.EINVAL => error.ArgumentsInvalid,
+                                os.EISDIR => error.IsDir,
+                                os.ELOOP => error.SymLinkLoop,
+                                os.EMFILE => error.ProcessFdQuotaExceeded,
+                                os.ENAMETOOLONG => error.NameTooLong,
+                                os.ENFILE => error.SystemFdQuotaExceeded,
+                                os.ENODEV => error.NoDevice,
+                                os.ENOENT => error.FileNotFound,
+                                os.ENOMEM => error.SystemResources,
+                                os.ENOSPC => error.NoSpaceLeft,
+                                os.ENOTDIR => error.NotDir,
+                                os.EOPNOTSUPP => error.FileLocksNotSupported,
+                                os.EOVERFLOW => error.FileTooBig,
+                                os.EPERM => error.AccessDenied,
+                                os.EWOULDBLOCK => error.WouldBlock,
+                                else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                            };
+                        } else {
+                            break :result @intCast(os.fd_t, completion.result);
+                        }
+                    };
+                    completion.callback(completion.context, completion, &result);
+                },
+            }
+        }
+    };
+
+    /// This union encodes the set of operations supported as well as their arguments.
+    const Operation = union(enum) {
+        accept: struct {
+            socket: os.fd_t,
+            address: os.sockaddr,
+            address_size: os.socklen_t,
+            flags: u32,
+        },
+        close: struct {
+            fd: os.fd_t,
+        },
+        openat: struct {
+            fd: os.fd_t,
+            path: [*:0]const u8,
+            flags: u32,
+            mode: os.mode_t,
+        },
+    };
+
```

Comment:
This could almost be a code generator or DSL! Such a terse description of everything supported!

Reply:
This is pretty similar to what my [wayland scanner](https://github.com/ifreund/zig-wayland) creates from protocol xml. Zig is lovely for stuff like this :)